src/hotspot/share/opto/addnode.cpp

```diff
@@ -1004,14 +991,29 @@ const Type *XorINode::add_ring( const Type *t0, const Type *t1 ) const {
   const TypeInt *r0 = t0->is_int(); // Handy access
   const TypeInt *r1 = t1->is_int();
 
-  // Complementing a boolean?
-  if( r0 == TypeInt::BOOL && ( r1 == TypeInt::ONE
-                               || r1 == TypeInt::BOOL))
-    return TypeInt::BOOL;
+  if( !r0->is_con() || !r1->is_con() ) {
+    //not a constant
 
-  if( !r0->is_con() || !r1->is_con() ) // Not constants
-    return TypeInt::INT;        // Any integer, but still no symbols.
+    // result of xor can only have bits sets where any of the
+    // inputs have bits set. lo can always become 0.
+
+    if ((r0->_lo >= 0) &&
+        (r0->_hi > 0)  &&
+        (r1->_lo >= 0) &&
+        (r1->_hi > 0)) {
+      // hi - set all bits below the highest bit. Using round_down to avoid overflow.
+      const TypeInt* t1x = TypeInt::make(0, round_down_power_of_2(r0->_hi) + (round_down_power_of_2(r0->_hi) - 1), r0->_widen);
+      const TypeInt* t2x = TypeInt::make(0, round_down_power_of_2(r1->_hi) + (round_down_power_of_2(r1->_hi) - 1), r1->_widen);
+      return t1x->meet(t2x);
+    }
```

Comment:
I know you only moved this code - but can you add some more comments about what it does? Maybe add some examples? It would help me review this code faster if I don't have to figure it out myself ;)